### DISCUSSION TITLE
(BFU)TextField

### DISCUSSION THREAD

**Discussion by RAN-Vestjylland** [BlazorFluentUI/BlazorFluentUI#280] [Ideas]
Hi all :-)

I have previously made an issue about binding an integer to the BFUTextField in v4.0.2. It is not possible at all. So my concern is that for this to work the best way there needs to be made either a generic component where TItem can be defined as System.Int16/32/64, System.Double and so on to support binding all simple datatypes other than strings. Or the other way, make a specific NumberInput component that is not a TextField (Either with or without BFU prefix.)

Hope my concerns makes sense to all :-) Thank you :-)

**Comment by vnbaaij** [comment]
How about changing it like this:
```
 public class TextField : TextFieldBase<string> { }

 public partial class TextFieldBase<TValue> : FluentUIComponentBase
```
So every `TextField` already in use in demo sites and elsewhere stays exactly the same when dealing with `string` values and you can use `TextFieldBase<int>` etc when you want to use a numerical value (together with the `InputType=InputType.Number` parameter)

Would that be workable solution?

**Reply by RAN-Vestjylland** [reply]
I'm back from Easter vacation.

I see you have been working hard on a fix. It looks really good and I think your generic solution would work very well.

**Reply by vnbaaij** [reply]
Hi, new version is even better! Checking it in later this morning. Added a new 'overload' for the number variants which I called NumericTextField where you don't need to specify InputType and TValue (when bound to a value). What do you think about the name? Or would TextFieldNumber or TextFieldNumber be better/clearer?

**Reply by RAN-Vestjylland** [reply]
I'm not sure how you define the generic in the razor part. But if it's a TValue attribute, then the name could be GenericTextField so you have to define the TValue with an enum TextFieldType.Integer, Decimal, Float etc.

But if you make them all specific for the type so the type is defined in the overloaded code, then PreFix naming would sound better I think. IntegerInputField maybe?? DecimalInputField, FloatInputField.

**Reply by vnbaaij** [reply]
It is a TValue parameter indeed, but it only accepts numeric types (int, long, short, float, double, and decimal) and throws an error with other types. This works the same as with the native InputNumber component hence I think it makes most sense to follow that naming 'convention'.

We don't need an extra enum for the types (only makes it more complicated internally to get the actual type (enum->map to string->get the actual type). Specifying the type with the TValue parameter (i.e. `TValue="int"`, etc.) is standard way of supplying that info to components already. If the value is bound to a variable (with the `@bound-Value=...` syntax, the type will be inferred from that variable and you don't need to supply `TValue` parameter at all.

**Reply by RAN-Vestjylland** [reply]
Perfect. I was not sure on how you actually meant to define that. So all is good I believe :-) Naming convention sounds perfect. I really don't have any attitude towards that :-) I'm just happy I don't have to deal with the issues I have now, with a new version.

Oh... another question that just popped into mind: Is it still possible to type in letters etc. into a numeric field or are they completely limited for numbers and dot/comma (depending on countries) for decimal values?

Do you have an idea when v5 will be released as a nuget package with the new numeric components?

**Comment by vnbaaij** [comment]
The changes are now live on the demo site. Test NuGet packages for 5.1 are now available on the nightly feed. I'm installing them in my test site now and if I don't find any strange things, I'll push them to NuGet later today.

And, yes the numeric fields are completely limited to numbers.

**Reply by RAN-Vestjylland** [reply]
Everything looks amazing. But there may still be a little issue with something. (I hope the attached image is working!)

![2021-04-06 10_19_52-Blazor Fluent UI Webassembly Demo](https://user-images.githubusercontent.com/78351622/113680949-b8e80500-96c1-11eb-9c24-134b01581da9.png)

**Reply by vnbaaij** [reply]
Yes, Ive noticed that too. But is seems the native InputNumber gives the same validation errors on float and double. I think it has to do with the input format. Will look at it some more, but will not hold back the release on it.

**Reply by RAN-Vestjylland** [reply]
Perfect. I can't wait to use the released components :-) Thank you for all the good work you have done!

**Reply by vnbaaij** [reply]
Packages are on NuGet now!

**Reply by RAN-Vestjylland** [reply]
I may be a little slow here. But is it the same package for 5.1 as the v4?
https://www.nuget.org/packages/BlazorFluentUI.AllComponents/

**Reply by vnbaaij** [reply]
No! You need BlazorFluentUI.CoreComponents and BlazorFluentUI.ListComponents. The rest is old and will be de listed soon.

**Reply by RAN-Vestjylland** [reply]
I updated to the new version. After removing all the BFU prefix it only have one minor issue that the webpage does not cover. Link component gives the following error:
![2021-04-06 13_35_49-](https://user-images.githubusercontent.com/78351622/113704991-2d7c6d00-96dd-11eb-92aa-a596ca0bc2a1.png)

I'm not sure how to solve this because it expects a method call inside the ChildContent property on the Link component.

**Comment by vnbaaij** [comment]
Do you have a `@using BlazorFluentUI` in your _Imports.razor?
If so, try using the full component name: <BlazorFluentUI.Link>...</BlazorFluentUI.Link>

**Reply by RAN-Vestjylland** [reply]
I do have @using BlazorFluentUI in my _Imports.razor of course. But it works fine with <BlazorFluentUI.Link ...>Something</BlazorFluentUI.Link>.

Maybe it's a tiny little bug :-) No rush!

**Reply by vnbaaij** [reply]
No, it is not a bug. The razor rendering in VS gets confused because it thinks you are using a HTML link

**Reply by RAN-Vestjylland** [reply]
ok.

Something else happens: 
![2021-04-06 14_04_05-VES TBS3 (Running) - Microsoft Visual Studio](https://user-images.githubusercontent.com/78351622/113708070-150e5180-96e1-11eb-82a6-722bda3a7e5c.png)

When I do this:
`<BlazorFluentUI.Link Href="/PrisBeregning/">Pris beregning</BlazorFluentUI.Link>`

**Reply by vnbaaij** [reply]
You are referring to the V4 script (BFUBaseComponent). 
Link to the V5 script with  `<script src="_content/BlazorFluentUI.CoreComponents/blazorFluentUI.js"></script>`

**Reply by RAN-Vestjylland** [reply]
Ohhhhhh you are a sharp person. Thank you!

**Reply by vnbaaij** [reply]
Can you share link or screenshot(s) from the site you are using it in? Curious to what it looks like in a real-life situation.

**Reply by RAN-Vestjylland** [reply]
I'm not sure i'm allowed to share that much information or show that much. But I can tell you that all is working with the Link component. Now i am running into a problem wiht the Dropdown component having a null pointer exception on SetStyle().

![2021-04-06 14_29_11-VES TBS3 (Running) - Microsoft Visual Studio](https://user-images.githubusercontent.com/78351622/113710845-800d5780-96e4-11eb-8457-24a8611a1a69.png)

**Reply by RAN-Vestjylland** [reply]
I thought it was "enough" to just apply it empty as well like this:
`<Dropdown Label="Skabelon" ItemsSource="@Model.SchemeOptions" @bind-SelectedOption="Model.SelectedSchemeOption" Style="" />`
It still gives the SetStyle() null pointer exception though.

**Reply by vnbaaij** [reply]
Ok, no problem. As I said, just curious.

`SetStyle` is independent from `Style` parameter. Example in my local demo site (server version) is working without `Style`, so must be something else. No indication on what that could be though.
![image](https://user-images.githubusercontent.com/1761079/113713152-4722b200-96e7-11eb-9274-b2bfbe25c0f9.png)
